brml/swap/src/lib.rs

```diff
@@ -1,1022 +1,1400 @@
+			});
+
+			Self::deposit_event(RawEvent::SwapTokenSuccess(token_amount_in, token_amount_out));
+
+			Ok(())
+		}
+
+		/// User claims bonus from only one pool
+		#[weight = T::DbWeight::get().reads_writes(1, 1)]
+		pub fn claim_bonus(
+			origin,
+			pool_id: T::PoolId
+		) -> DispatchResult {
+			let claimer = ensure_signed(origin)?;
+			ensure!(Pools::<T>::contains_key(pool_id), Error::<T>::PoolNotExist);
+			ensure!(Pools::<T>::get(pool_id).active, Error::<T>::PoolNotActive);
+
+			// ensure the user has pool tokens for the pool
+			ensure!(UserPoolTokensInPool::<T>::contains_key(&claimer, pool_id), Error::<T>::UserNotInThePool);
+
+			Self::update_unclaimed_bonus_related_states(&claimer, pool_id)?;
+
+			UserUnclaimedBonusInPool::<T>::mutate(&claimer, pool_id, |(unclaimed_bonus_balance, _block_num)| {
+				// issue corresponding BNC bonus to the user's account
+				T::AssetTrait::asset_issue(T::BNCAssetId::get(), &claimer, *unclaimed_bonus_balance);
+				// mutate the user's unclaimed BNC bonus to zero
+				*unclaimed_bonus_balance = Zero::zero();
+			});
+
+			Ok(())
+		}
+
+		/// ******************************************************
+		/// ***  Above are the exchange functions.			  ***
+		/// ***  Below are the exchange management functions. ***
+		/// ******************************************************
+		#[weight = T::DbWeight::get().reads_writes(1, 1)]
+		pub fn create_pool(
+			origin,
+			swap_fee_rate: T::Fee,  // this number is an integer to avoid precision loss, should be divided by fee precision constant when used.
+			token_for_pool_vec: Vec<PoolCreateTokenDetails<T>>,
+		) -> DispatchResult {
+			let creator = ensure_signed(origin)?;
+
+			// swap fee rate should be greater or equals to MinimumSwapFee.
+			ensure!(swap_fee_rate >= T::MinimumSwapFee::get(), Error::<T>::FeeRateExceedMinimumLimit);
+			// swap fee rate should be greater or equals to MaximumSwapFee.
+			ensure!(swap_fee_rate <= T::MaximumSwapFee::get(), Error::<T>::FeeRateExceedMaximumLimit);
+
+			// create three iterators for the map to be able to use multiple times.
+			let map_iter = token_for_pool_vec.iter();
+			ensure!(map_iter.len() <= T::NumberOfSupportedTokens::get() as usize, Error::<T>::TooManyTokensToPool);
+
+			let mut total_weight = T::PoolWeight::from(0);
+
+			let map_iter_1 = token_for_pool_vec.iter();
+			// ensure all the elements of the tokenForPoolMap are ok.
+			for token_info in map_iter_1 {
+				ensure!(T::AssetTrait::token_exists(token_info.token_id), Error::<T>::TokenNotExist);
+				ensure!(token_info.token_balance > Zero::zero(), Error::<T>::AmountBelowZero);
+
+				// get the user's balance for a specific token
+				let user_token_balance = T::AssetTrait::get_account_asset(token_info.token_id, &creator).balance;
+				ensure!(user_token_balance >= token_info.token_balance, Error::<T>::NotEnoughBalance);
+				// Add up the total weight
+				total_weight += token_info.token_weight;

```

Comment:
'+' shoule be replaced by saturating_add

Reply:
fixed.